opnode/rollup/driver/step.go

```diff
@@ -283,6 +291,32 @@ func (d *outputImpl) insertHeadBlock(ctx context.Context, fc l2.ForkchoiceState,
 	if err != nil {
 		return nil, fmt.Errorf("failed to get execution payload: %w", err)
 	}
+	// Sanity check payload before inserting it
+	if len(payload.TransactionsField) == 0 {
+		return nil, errors.New("no transactions in returned payload")
+	}
+	txns := payload.Transactions()
+	if txns[0].Type() != types.DepositTxType {
+		return nil, fmt.Errorf("first transaction was not deposit tx. Got %v", txns[0].Type())
+	}
+	// Ensure that the deposits are first
+	var lastDeposit int
+	doneWithDeposits := false
+	for i, tx := range txns {
+		if tx.Type() == types.DepositTxType && doneWithDeposits {
+			d.log.Error("Produced an invalid block where the deposit txns are not all at the start of the block", "tx_idx", i, "tx", tx.Hash(), "lastDeposit", lastDeposit)
+			return nil, fmt.Errorf("deposit tx (%d:%s) after other tx in l2 block with prev deposit at idx %d", i, tx.Hash(), lastDeposit)
+		} else if tx.Type() == types.DepositTxType {
+			lastDeposit = i
+		} else {
+			doneWithDeposits = true
+		}
+	}
+	// TODO: Is this an error? This would likely happen on the verifier as well which means a chain halt.
```

Comment:
not all transactions may be deposits right?

Reply:
Right, this is most important on the sequencer side when creating an unsafe block. When inserting a safe block, it could have dropped a transaction. In general, I want to know when there is a mismatch between the attributes and payload